lucet-runtime/lucet-runtime-internals/src/instance.rs

```diff
@@ -565,6 +565,37 @@ impl Instance {
                 "entrypoint function cannot be null; this is probably a malformed module",
             ));
         }
+
+        let fn_id: u32 = self
+            .module
+            .function_manifest()
+            .iter()
+            .enumerate()
+            .find(|(_, fn_spec)| fn_spec.relative_addr(func as u64) == Some(0))
+            .map(|(fn_id, _)| fn_id as u32)
+            .ok_or(Error::InvalidArgument(
+                "entrypoint function must be the start of a function declared in this module",
+            ))?;
```

Comment:
This is correct code, and I'm impressed by it, but rather than lookup into the module from the function pointer, lets instead have Module always return this metadata alongside the function pointer.

Can you have `Module::get_export_func`, `get_start_func`, and `get_func_from_idx` return structs that have both the function pointer, and a ref to the signature, inside it?

Reply:
I think having those functions return a pointer/signature pair ends up complicating downstream interfaces in a bad way.. I'm not as opposed to making all uses of internal functions take a struct like `FunctionHandle(*const extern "C" fn(), &Signature)` but that still obligates users to look up the signature when they might not need it, as well as becoming a change to `lucet_vmctx.h` to return structs or tuples or whatever (though that might be fine?)

Since signature lookup will end up happening by matching function pointers anyway[1], how's taking this chain and putting it behind a function like `Module::signature_for_func` sound? 

[1] Looking up by function pointers would be necessary to retrieve signatures for functions obtained by table index, where the space of `(table_idx, fn_idx)` doesn't very nicely to the space of functions with signatures in a Module. Not matching function pointers would mean tables have to have an index to the internal function space and adds more indirection for indirect calls :(

Reply:
I don't know of any uses of looking up a function that don't also require the signature eventually. And I'd ideally like to change Instance::run to take a function handle in substitute for the export name, and make users look up the function from a `Module::get_export_function` interface.

For the table lookup, we should add a module-data that maps table index to function index - it can just be a vector for now. We'll need to expand that representation when we implement table initializers that compute offsets from import globals. So, you can put the simple version of that in now.

I don't think its a problem to change the C API. We don't expect any of the runtime APIs to be stable as we roll out the changes required for spec compliance. We left out a lot of this complexity in order to have simpler interfaces that were easy to iterate on, but we're going to have to add some indirection.

Reply:
btw I don't think you need to change Instance::run and expose Module::get_export_function in this PR. but it may be a good thing to think about. It will require some way of checking that the export function passed to an instance belongs to the module of the instance. I'm not sure if you can use lifetimes to do that in the Rust type system or not, but for the C API itll require some sort of runtime check.

Reply:
A mapping from table index to function index makes sense and cleans up that concern well.

I thought there was a cases where we would end up grabbing the signature, discard it, and look it up again to check, but on reread that's not actually a concern because we'd be passing the handle including signature along. So I'm totally on board :)